server.py

```diff
@@ -1,62 +1,84 @@
+        self.socket_port = socket_port
+        self.refresh_rate = refresh_rate
+        self.ws_server = None
+
+    async def start(self):
+        self.log.info('Starting the websocket server')
+
+        def stop_cb():
+            return asyncio.create_task(self.stop())
 
-async def sendLocations(websocket, path):
-    while True:
-        output = json.dumps(getLocations())
         try:
-            await websocket.send(output)
-        except (websockets.exceptions.ConnectionClosedOK, websockets.exceptions.ConnectionClosedError, websockets.exceptions.ConnectionClosed):
-            await websocket.wait_closed()
-            logging.info("Client connection closed")
-            break
-        finally:
-            await asyncio.sleep(REFRESH_RATE)
+            asyncio.get_running_loop().add_signal_handler(signal.SIGINT, stop_cb)
+            asyncio.get_running_loop().add_signal_handler(signal.SIGTERM, stop_cb)
+        except NotImplementedError as e:
+            log.exception(f'Not implemented')
+
+        self.ws_server = await websockets.serve(self.send_locations, self.socket_host, self.socket_port)
+
+        await self.ws_server.wait_closed()
+
+        self.log.info('Websocket server closed')
+
+    async def stop(self):
+        self.ws_server.close()
+        self.minecraft.disconnect()
 
-def getLocations():
-    output = []
-    response = mcr.command("/list")
-    match = re.search(r"([0-9]+).+:(.+)", response)
-    numberOfPlayers = int(match.group(1))
+    async def send_locations(self, websocket, path):
+        while True:
+            players = self.minecraft.get_players()
+            player_data = json.dumps([asdict(p) for p in players])
+            self.log.debug(f'Sending {player_data}')
+            await websocket.send(player_data)
+            await asyncio.sleep(self.refresh_rate)
 
-    if numberOfPlayers != 0:
-        playerList = match.group(2).strip().split(", ")
-        for playerName in playerList:
-            response = mcr.command("/data get entity {} Pos".format(playerName))
-            match = re.search(r"(\[.*\])", response)
-            playerPosition = eval(match.group(1).replace("d",""))
 
-            response = mcr.command("/data get entity {} Dimension".format(playerName))
-            match = re.search(r"data: (.*)", response)
-            playerDimension = match.group(1).strip("\"")
+async def main(rcon_host, rcon_password, rcon_port, socket_host, socket_port, refresh_rate, log_level):
+    logging.basicConfig(level=log_level)
+    minecraft = MinecraftServer(rcon_host, rcon_password, rcon_port)
+    ws = WebSocketServer(minecraft, socket_host, socket_port, refresh_rate)
 
-            output.append({"name": playerName, "position": {"x": playerPosition[0], "y": playerPosition[1], "z": playerPosition[2]}, "dimension": playerDimension})
+    await ws.start()
 
-    return output
 
-def exitGracefully():
-    logging.warning("Shutting down")
-    sys.exit(1)
+if __name__ == '__main__':
```

Comment:
So my context clues as I've gone through this branch lead me to believe this is determining if this is the file being executed rather than being imported. Thus allowing the ability to test a specific script rather than having to instantiate the entire chain.